### PR TITLE
added the option to name docker container for functions and reuse them.

### PR DESCRIPTION
**Pull request by TheBelgarion**
## Description
currently crossplane beta render will always create new container on each run of beta render and only has the option to Remove, Stop, Orphan the container but will still create new container on each run.

the new annotation parameter 'render.crossplane.io/runtime-docker-name' tells beta render to create the container with a specific name and it will check if a container with that name already exists and even tries to restart it if it not running.
this greatly improves the performance then running beta render multiple times, eg. testing multiple crds.

#### Speed Comparison Test:

Setup:
* crossplane beta render xxx && crossplane beta validate xxx
* build crank image
* Tests Case setup
  * all tests done with preloaded images
  * done on mac with docker & podman with docker Compatibility
  * test set contains 8 crd creating 10 resources
  * running with 7 different functions

default:
* with no render.crossplane.io/runtime-docker-* settings
* creates and destroy container with random names for each render run
* Result:
  * tests took 55 seconds
  * 0 running or halted containers with functions

orphan:
* with render.crossplane.io/runtime-docker-cleanup: Orphan
* creates container with random names for each render run, does not destroy any container
* Result: 
  * tests took 45 seconds
  * 63 running containers with functions

stop:
* with render.crossplane.io/runtime-docker-cleanup: Stop
* creates container with random names for each render run, and stops them
* Result: 
  * tests took 45 seconds
  * 63 stopped containers with functions

named+stop:
* render.crossplane.io/runtime-docker-name: "\<name\>"
* render.crossplane.io/runtime-docker-cleanup: Stop
* creates container with fixed names on the first run, stop, starts and reuses them
* Result: 
  * tests took 45 seconds
  * 0 running named containers with functions

named+remove:
* render.crossplane.io/runtime-docker-name: "\<name\>"
* render.crossplane.io/runtime-docker-cleanup: Remove
* creates container with fixed names on  run and deletes the container
* Result: 
  * tests took 53 seconds
  *  0 running or halted containers with functions

named+orphan:
* render.crossplane.io/runtime-docker-name: "\<name\>"
* render.crossplane.io/runtime-docker-cleanup: Orphan
* creates container with fixed names on the first run and reuses them
* Result: 
  * tests took **24 seconds**
  * 7 running named containers with functions

this leads to an  improved render speed and speeds up the test by at over 50%,  which will be even higher the more crd you are testing.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

**Comment by bobh66** [comment]
@TheBelgarion thanks for the PR.  Can you provide some context around the scenario and the problem you are trying to solve?  It's not clean from the code changes exactly what the scenario is.  Thanks

**Comment by TheBelgarion** [comment]
@bobh66 
I updated description and added a speed comparison test with a small test set.

**Comment by bobh66** [comment]
@TheBelgarion thanks!  What happens if you set:

```
render.crossplane.io/runtime-docker-name: "<name>"
render.crossplane.io/runtime-docker-cleanup: Stop
```

Do you get a similar performance improvement as the last case but leave 7 containers stopped instead of running?  It seems like that might be the optimal case where you get the benefit of the pre-existing containers that only need to be restarted but you also don't have the containers left running at the end.

**Comment by TheBelgarion** [comment]
@bobh66 
updated description.
starting & stopping does seem to add quite a bit and reduces the effectiveness but it works as well.

**Comment by bobh66** [comment]
@TheBelgarion thanks, that's interesting.  Have you tried using Pause instead of Stop?

**Comment by TheBelgarion** [comment]
@bobh66 
did add any additional docker runtime calls but guess it still slower because it will always wait til the container is ready.

File: cmd/crank/beta/render/cmd.go
Comment (negz) [inline cmd/crank/beta/render/cmd.go:87]: It's a bit long, but maybe `runtime-docker-container-name` would make it super clear?

File: cmd/crank/beta/render/runtime_docker.go
Comment (negz) [inline cmd/crank/beta/render/runtime_docker.go:184]: I'd prefer to return early here to avoid nesting the bulk of the method's logic inside this conditional.

https://github.com/crossplane/crossplane/tree/master/contributing#return-early

File: cmd/crank/beta/render/runtime_docker.go
Comment (negz) [inline cmd/crank/beta/render/runtime_docker.go:193]: Similarly here - I believe returning early would keep this method less nested.

File: cmd/crank/beta/render/runtime_docker.go
Comment (negz) [inline cmd/crank/beta/render/runtime_docker.go:287]: Why pass `r.Name` as an argument? `r.findContainer` is a method on `r`, so it can already access `r.Name`.

It seems like the `startContainer` argument is always `true`, right? If so I'd suggest removing it.

File: cmd/crank/beta/render/runtime_docker.go
Comment (negz) [inline cmd/crank/beta/render/runtime_docker.go:173]: Nit: Please use the same variable name as `Start`. When I see `c` and `cli` in closely related functions it makes me think they're different things.

```suggestion
func (r *RuntimeDocker) findContainer(ctx context.Context, c *client.Client, containerName string, startContainer bool) (string, string) {
```

**Review by negz — COMMENTED** [review]
Thanks! This seems useful - I wonder if reusing containers when possible should be the default behavior?

File: cmd/crank/beta/render/cmd.go
Comment (TheBelgarion) [reply cmd/crank/beta/render/cmd.go:87]: i did not want to changed the default behavior also i wanted to have flexibility and not just name the container like the image.

also i think runtime-docker-name also enough also in the scheme of runtime-docker-image, <imho>

File: cmd/crank/beta/render/runtime_docker.go
Comment (TheBelgarion) [reply cmd/crank/beta/render/runtime_docker.go:184]: done

File: cmd/crank/beta/render/runtime_docker.go
Comment (TheBelgarion) [reply cmd/crank/beta/render/runtime_docker.go:193]: dont see it here, its a loop ?

File: cmd/crank/beta/render/runtime_docker.go
Comment (TheBelgarion) [reply cmd/crank/beta/render/runtime_docker.go:287]: done

File: cmd/crank/beta/render/runtime_docker.go
Comment (TheBelgarion) [reply cmd/crank/beta/render/runtime_docker.go:173]: i hate one letter variable :-) 
done

Comment (TheBelgarion) [reply cmd/crank/beta/render/runtime_docker.go:173]: fixed it to be cli everythere

File: cmd/crank/beta/render/cmd.go
Comment (TheBelgarion) [reply cmd/crank/beta/render/cmd.go:87]: we could add if no name is given use the image name as a default name instead of empty one. But it could also break thing, e.g. you do that one some system where you have also the named once which would possible lead to them beeing deleted. which is not what someone might want.

also i will probably update my test script so in the last render it just uses another function.yaml which deletes the container after for cleanup.

**Comment by TheBelgarion** [comment]
@bobh66  @negz  can you check now - i even did the merge conflict with current base.
@negz next time use git mv instead of just moving folders :-)

File: cmd/crank/render/runtime_docker.go
Comment (negz) [inline cmd/crank/render/runtime_docker.go:192]: Nit: `Container` is an unusual variable name for Go. I'd recommend `c`.

https://github.com/crossplane/crossplane/tree/main/contributing#use-descriptive-variable-names-sparingly

**Review by negz — APPROVED** [review]
Please address the one nit, but otherwise this looks good.

**Comment by TheBelgarion** [comment]
@negz  fixed and signed off :-)

**Comment by jbw976** [comment]
@TheBelgarion was asking about the [failing Trivy scan check](https://github.com/crossplane/crossplane/actions/runs/11549953254/job/32145405844?pr=5878) on this PR. I have seen it fail semi-frequently recently due to rate limiting / throttling, but I also have seen failures like this one that don't have any clear error message. That may be something we should look into, but since this PR does not introduce any new dependencies, it shouldn't block this PR at all.

In the meantime, @TheBelgarion it looks like a merge conflict now exists here, would you mind resolving that? Feel free to ping me directly after that's done and we'll get this merged ✅ - thank you!

**Comment by gjbravi** [comment]
Thanks @TheBelgarion, huge improvement!

We will test the new binary in our unit-tests CI.

**Pull request merged**
